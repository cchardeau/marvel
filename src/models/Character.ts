export type TCharacter = {
  id: number
  name: string
  thumbnail: string
  description: string
  comicsCount: number
  comics: string[]
}
