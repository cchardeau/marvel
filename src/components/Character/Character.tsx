import React from 'react'
import Button from '@material-ui/core/Button'
import Card from '@material-ui/core/Card'
import CardActions from '@material-ui/core/CardActions'
import CardContent from '@material-ui/core/CardContent'
import CardMedia from '@material-ui/core/CardMedia'
import Grid from '@material-ui/core/Grid'
import { makeStyles } from '@material-ui/core/styles'

import CharacterDetails from '../CharacterDetails/CharacterDetails'
import { TCharacter } from '../../models/Character'

type Props = {
  character: TCharacter
}

const useStyles = makeStyles((theme) => ({
  cardGrid: {
    paddingTop: theme.spacing(8),
    paddingBottom: theme.spacing(8)
  },
  card: {
    height: '100%',
    display: 'flex',
    flexDirection: 'column'
  },
  cardMedia: {
    paddingTop: '56.25%' // 16:9
  },
  cardContent: {
    flexGrow: 1
  }
}))

const Character: React.FC<Props> = ({ character }) => {
  // styling
  const classes = useStyles()

  return (
    <React.Fragment key={character.id}>
      <Grid item xs={12} sm={6} md={4}>
        <Card className={classes.card}>
          <CardMedia
            className={classes.cardMedia}
            image={character.thumbnail}
            title={`${character.name} image`}
          />
          <CardContent className={classes.cardContent}>
            <CharacterDetails character={character} />
          </CardContent>
          <CardActions>
            <Button size="small" color="primary">
              Set as favorite
            </Button>
          </CardActions>
        </Card>
      </Grid>
    </React.Fragment>
  )
}

export default Character