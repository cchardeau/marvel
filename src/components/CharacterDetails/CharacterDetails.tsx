import React, { useState } from 'react'
import Typography from '@material-ui/core/Typography'
import Dialog from '@material-ui/core/Dialog'
import DialogTitle from '@material-ui/core/DialogTitle'
import DialogContent from '@material-ui/core/DialogContent'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableRow from '@material-ui/core/TableRow'
import TableCell from '@material-ui/core/TableCell'
import { makeStyles } from '@material-ui/core/styles'

import { TCharacter } from '../../models/Character'

type Props = {
  character: TCharacter
}

const useStyles = makeStyles(() => ({
  characterName: {
    cursor: 'pointer'
  },
  dialogContent: {
    padding: 0
  }
}))

const CharacterDetails: React.FC<Props> = ({ character }) => {
  // init
  const [openDetails, setOpenDetails] = useState(false)

  // styling
  const classes = useStyles()
  
  const handleDetailsOpen = () => {
    setOpenDetails(true)
  }

  const handleDetailsClose = () => {
    setOpenDetails(false)
  }

  return (
    <React.Fragment key={`${character.id}-details`}>
      <Typography gutterBottom className={classes.characterName} variant="h5" component="h2" onClick={handleDetailsOpen}>
        {character.name}
      </Typography>
      <Dialog onClose={handleDetailsClose} aria-labelledby="de-dialog" open={openDetails}>
        <DialogTitle>
          {character.name} - details
        </DialogTitle>
        <DialogContent className={classes.dialogContent} dividers>
          <Table>
            <TableBody>
              <TableRow>
                <TableCell align="center">
                  <Typography><strong>Name</strong></Typography>
                </TableCell>
                <TableCell align="center">
                  <Typography>{character.name}</Typography>
                </TableCell>
              </TableRow>
              <TableRow>
                <TableCell align="center">
                  <Typography><strong>Description</strong></Typography>
                </TableCell>
                <TableCell align="center">
                  <Typography>{character.description}</Typography>
                </TableCell>
              </TableRow>
              <TableRow>
                <TableCell align="center">
                  <Typography><strong>Image</strong></Typography>
                </TableCell>
                <TableCell align="center">
                  <img src={character.thumbnail} alt={`${character.name} detail`} />
                </TableCell>
              </TableRow>
              <TableRow>
                <TableCell align="center">
                  <Typography><strong>Number of occurrences</strong></Typography>
                </TableCell>
                <TableCell align="center">
                  <Typography>{character.comicsCount}</Typography>
                </TableCell>
              </TableRow>
              <TableRow>
                <TableCell align="center">
                  <Typography><strong>First occurrences</strong></Typography>
                </TableCell>
                <TableCell align="center">
                  {character.comics.map((comic) => (<Typography key={`${character.name}-${comic}`}>{comic}</Typography>))}
                </TableCell>
              </TableRow>
            </TableBody>
          </Table>
        </DialogContent>
      </Dialog>
    </React.Fragment>
  )
}

export default CharacterDetails