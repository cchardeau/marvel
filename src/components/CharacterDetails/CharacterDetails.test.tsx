import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import CharacterDetails from './CharacterDetails'
import { TCharacter } from '../../models/Character'

const mockCharacter: TCharacter = {
  id: 1,
  name: 'name',
  thumbnail: 'https://thumbnail.jpg',
  description: 'description',
  comicsCount: 42,
  comics: []
}

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<CharacterDetails character={mockCharacter} />, div)
  ReactDOM.unmountComponentAtNode(div)
})

it('renders correctly', () => {
  const output = renderer.create(<CharacterDetails character={mockCharacter} />).toJSON()
  expect(output).toMatchSnapshot()
})
