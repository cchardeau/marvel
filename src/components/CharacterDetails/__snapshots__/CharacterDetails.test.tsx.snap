// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<h2
  className="MuiTypography-root makeStyles-characterName-48 MuiTypography-h5 MuiTypography-gutterBottom"
  onClick={[Function]}
>
  name
</h2>
`;